src/tasks.ts

```diff
@@ -96,59 +114,120 @@ export async function main(): Promise<void> {
+        sourceEpoch,
+        targetEpoch,
+        tree,
+    };
+    const filename = saveArtifact(artifact);
+    debug(`Artifact saved to ${filename}`);
+
+    // Uploading the tree optimistically
+    debug('Uploading the artifact to IPFS');
+    const cid = await uploadFile(filename);
+    debug('Tree uploaded, CID is', cid.IpfsHash);
+
+    const report = {
+        consensusVersion: Shared.CONSENSUS_VERSION,
+        refSlot: refSlot,
+        treeRoot: tree.root,
+        treeCid: cid.IpfsHash,
+        distributed: distributed,
+    };
+    const reportHash = hashReport(report);
+
+    const { currentFrameMemberReport } =
+        await Shared.HASHCONSENSUS.getConsensusStateForMember(
+            Shared.SIGNER.address,
+        );
+    if (currentFrameMemberReport == reportHash) {
+        debug('Provided hash already submitted');
+        return;
+    }
+
+    debug('Simulating sending report hash');
+    await Shared.HASHCONSENSUS.connect(Shared.SIGNER).submitReport.staticCall(
+        refSlot,
+        reportHash,
+        Shared.CONSENSUS_VERSION,
     );
     debug('Simulation successfull!');
 
-    debug('Uploading the tree to IPFS');
-    const cid = await uploadTree(tree);
-    debug('Tree uploaded, CID is', cid.toString());
-
     debug('Sending the report');
-    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
-        targetEpoch,
-        tree.root,
-        distributed,
-        cid.toString(),
+    await Shared.HASHCONSENSUS.connect(Shared.SIGNER).submitReport(
+        refSlot,
+        reportHash,
+        Shared.CONSENSUS_VERSION,
     );
-
     debug('Report sent, storing the artifact');
-    saveArtifact(from, to, tree);
+
+    const { currentFrameConsensusReport } =
+        await Shared.HASHCONSENSUS.getConsensusStateForMember(
+            Shared.SIGNER.address,
+        );
+
+    if (currentFrameConsensusReport == ZeroHash) {
+        debug('No consensus reached');
+        return;
+    }
+
+    if (currentFrameConsensusReport != reportHash) {
+        debug('Oracle`s hash differs from consensus report hash. Exiting');
+        return;
+    }
+
+    // TODO: check for fast lane
+    const treeRoot = await Shared.ORACLE.treeRoot();
+    if (treeRoot == tree.root) {
+        debug('Report already settled');
+        return;
+    }
+
+    debug('Settling the report');
+    debug('Simulating sending report data');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReportData.staticCall(
+        report,
+        Shared.CONTRACT_VERSION,
+    );
+    debug('Simulation successfull!');
+
+    debug('Sending report data');
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReportData(
+        report,
+        Shared.CONTRACT_VERSION,
+    );
+    debug('Report data sent!');
+    debug('Done');
 }
 
-async function nextReportEpoch(blockTag: BlockTag) {
-    const r = await Shared.ORACLE.nextReportEpoch({
+async function nextReportRefSlot(blockTag: BlockTag): Promise<Slot> {
```

Comment:
Can we name it `currentReportRefSlot`. a bit misleading seeing `HASHCONSENSUS.getCurrentFrame` inside. Next or not next from the app perspective should be decided somewhere outside this func